src/main/java/org/opensearch/plugin/insights/rules/transport/top_queries/TransportTopQueriesAction.java

```diff
@@ -68,27 +75,143 @@ public TransportTopQueriesAction(
         this.queryInsightsService = queryInsightsService;
     }
 
+    ActionListener<TopQueriesResponse> createInMemoryDataCollectionListener(
```

Comment:
I see more scope of refactoring in this class as it contains a lot of logic. Transport classes should ideally be lean, just responsible to route the request, whereas any custom logic or response handling stuff should be decoupled from transport layer.

This is something you can think/refactor later on.